### ISSUE TITLE
Better array and object inferred types with checkJs and strict / noImplicitAny

### ISSUE THREAD

**Issue by cyrilletuzi** [Microsoft/TypeScript#16137]
**TypeScript Version:**  2.3.3

**Code**

In a JavaScript file (_not_ TypeScript), with the new `checkJs` option enabled : 

```js
let list = [`value 1`, `value 2`, 10];

let user = {
  name: `John`,
  age: 20
};

user.newDynamicProperty = `test`;
```

**Actual behavior:**

In normal mode, no errors are raised, and it's OK as it's the normal JS behavior.

In `strict` or `noImplicitAny` mode, no errors are raised either, despite the fact that the inferred type for `user` is : `{ [x: string]: any; name: string; age: number; }`. For the array, it's `(string | number)[]` but in the purpose of error checking it's like `any[]`.

**Expected behavior:**

In `strict` or `noImplicitAny` mode, errors should be raised.

It seems important to me for 2 reasons : 
- with the current behavior, manual type via JSDoc is required to have a correct error checking like in TS, so it requires more work than in TS (usually it should be less) and despite the fact that TS already knows how to manage that ;
- it's specially painful for objects as we have no interfaces in JS, and as multi-lines 
`@type` or `@param` JSDoc comments are not supported by TS yet (see #11597).

**Comment by mhegazy** [comment]
in a .js file there are few places where `any` is injected. this is one of them, but there are other ones. `--noImplicitAny` really is not honored here correctly. as a matter of fact we have an issue to disable reporting `--noImplicitany` errors in a .js files, since they are incomplete at best (see https://github.com/Microsoft/TypeScript/issues/14617).

The assumption here is js file checking with checkJs is a loose version of checking; our mental model of this feature is that there is a spectrum of strictness, where: 
checking .js file < checking .ts file < checking .ts file with `--strict`. 

Just for curiosity; have you considered moving your files to .ts?

**Comment by cyrilletuzi** [comment]
I use Angular so I'm used to TypeScript and I love it, but I'm very interested with the new `checkJs` for some other projects where no framework and no transpilation is needed, to stay in native JS but using `checkJs` to have the power of TS for type checking.

After more investigation, array behavior is quite OK : you can mix values of different types on array creation, but once the type is inferred, `checkJs` will raise errors if I don't respect it.

It's OK for objects with `@typedef` JSDoc, but it's quite a heavy solution for some small case. As it's just because `[x: string]: any;` is added to objects type, wouldn't it possible to just change that in strict mode, like does TS ?

I'll post another message in the other topic you mention.

**Comment by que-etc** [comment]
@cyrilletuzi 

May I ask you why you did you decide to close this issue?

Inferred type of an object still has `[x: string] any;`  and I can't seem to find a workaround for it, except for providing a `@typedef` for an every object being created.

It would be so much helpful if TypeScript raised errors when it encounters an undefined property, just like it happens with `ts` files:
```js
const sample = {
    DEFINED_PROPERTY: 'FOO'
}

sample.UNDEFINED_PROPERTY.indexOf('BAR');
```
